src/som/interpreter/StackUtils.java

```diff
@@ -0,0 +1,256 @@
+ * <p>
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ * <p>
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package som.interpreter;
+
+import com.oracle.truffle.api.frame.FrameSlot;
+import com.oracle.truffle.api.frame.FrameSlotTypeException;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.nodes.ExplodeLoop;
+
+import som.vm.Universe;
+import som.vmobjects.SAbstractObject;
+import som.vmobjects.SBlock;
+import som.vmobjects.SMethod;
+
+
+/**
+ * Arguments.
+ * 0) Interpreter
+ * 2) Invokable
+ * 3) Arguments
+ *
+ * Slots:
+ * 0) Execution stack
+ * 1) Stack pointer
+ * 2) onStack marker
+ */
+public class StackUtils {
+  private enum FrameArguments {
+    INTERPRETER,
+    INVOKABLE,
+    ARGUMENTS
+  }
+
+  public static void initializeStackSlots(final VirtualFrame frame,
+      final FrameSlot executionStackSlot,
+      final SMethod method) {
+
+    int stackLength = (int) method.getMaximumLengthOfStack();
+    SAbstractObject[] stack = new SAbstractObject[stackLength];
+    for (int i = 0; i < stackLength; i++) {
+      stack[i] = Universe.current().nilObject;
+    }
+
+    // TODO - getting the arguments from the frame to put them on stack...
+    // could likely do both when putting them in the stack in the first place
+    // or just don't put them in the frame's arguments, is is useful anyway?
+    final SAbstractObject[] arguments = getCurrentArguments(frame);
+    System.arraycopy(arguments, 0, stack, 0, arguments.length);
+
+    frame.setObject(executionStackSlot, stack);
+    resetStackPointer(frame, method);
+  }
+
+  public static FrameOnStackMarker initializeStackMarkerSlot(VirtualFrame frame,
+      FrameSlot frameOnStackMarkerSlot) {
+    FrameOnStackMarker marker = new FrameOnStackMarker();
+    frame.setObject(frameOnStackMarkerSlot, marker);
+
+    return marker;
+  }
+
+  /***
+   * CURRENT STATE OF THE FRAME.
+   */
+
+  public static SMethod getCurrentMethod(final VirtualFrame frame) {
+    return (SMethod) frame.getArguments()[FrameArguments.INVOKABLE.ordinal()];
+  }
+
+  public static SAbstractObject[] getCurrentArguments(VirtualFrame frame) {
+    return (SAbstractObject[]) frame.getArguments()[FrameArguments.ARGUMENTS.ordinal()];
+  }
+
+  public static SAbstractObject[] getCurrentStack(VirtualFrame frame)
+      throws FrameSlotTypeException {
+    return (SAbstractObject[]) frame.getObject(getCurrentMethod(frame).getStackSlot());
```

Comment:
Is there any way to avoid this? Can't the stack slot just be a global constant of sorts?

Reply:
This should be fixed by the new commits. The `FrameDescriptor` and the associated frame slots are now global.
See https://github.com/SOM-st/som-java/blob/9ebec48390a88d2687c7c5f77d6227af46a3919c/src/som/interpreter/StackUtils.java#L63